engine/runtime/src/main/java/org/enso/interpreter/node/typecheck/TypeCheckValueNode.java

```diff
@@ -0,0 +1,181 @@
+package org.enso.interpreter.node.typecheck;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import org.enso.interpreter.node.ExpressionNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedConstructor;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.util.CachingSupplier;
+
+/** A node and a factory for nodes performing type checks (including necessary conversions). */
+public final class TypeCheckValueNode extends Node {
+  private @Child AbstractTypeCheckNode check;
+
+  TypeCheckValueNode(AbstractTypeCheckNode check) {
+    assert check != null;
+    this.check = check;
+  }
+
+  /**
+   * Wraps expression node with additional type check.
+   *
+   * @param original the expression node
+   * @param check node performing type check or {@code null}
+   * @return wrapper around {@code original} or directly {@code original} if there is {@code null}
+   *     check
+   */
+  public static ExpressionNode wrap(ExpressionNode original, TypeCheckValueNode check) {
+    if (check == null) {
+      return original;
+    } else {
+      return new TypeCheckExpressionNode(original, check);
+    }
+  }
+
+  /**
+   * Executes check or conversion of the value.
+   *
+   * @param frame frame requesting the conversion
+   * @param value the value to convert
+   * @return {@code null} when the check isn't satisfied and conversion isn't possible or non-{@code
+   *     null} value that can be used as a result
+   */
+  public final Object handleCheckOrConversion(
+      VirtualFrame frame, Object value, ExpressionNode expr) {
+    var result = check.executeCheckOrConversion(frame, value, expr);
+    if (result == null) {
+      throw panicAtTheEnd(value);
+    }
+    return result;
+  }
+
+  /**
+   * Combines existing type checks into "all of" check.
+   *
+   * @param comment description of the check meaning
+   * @param checks existing type checks
+   * @return node the composed check or {@code null} if no check is needed
+   */
+  public static TypeCheckValueNode allOf(String comment, TypeCheckValueNode... checks) {
+    if (checks == null) {
+      return null;
+    }
+    var list = Arrays.asList(checks);
+    var flatten =
+        list.stream()
+            .filter(n -> n != null)
+            .map(n -> n.check)
+            .flatMap(
+                n ->
+                    n instanceof AllOfTypesCheckNode all
+                        ? Arrays.asList(all.getChecks()).stream()
+                        : Stream.of(n))
+            .toList();
```

Comment:
@JaroslavTulach I remember you were saying that Truffle does not handle stream processing well. I'm surprised that it is not behind the boundary

Reply:
It is a node construction. Node construction is always outside of PE. So technically it is _behind the boundary_.

Reply:
The idea is: `Node` structure (e.g. AST) represents the program. If one needs to modify the program, then that means the program cannot be running in PE mode (at full speed) and everything switches to interpreter.